[TF FE] Support Complex Tensors for Slice Operation

ysrastogi opened the pull request:
**Details:** Support Complex Tensors for Slice Operation

rkazants commented on src/frontends/tensorflow_common/src/helper_transforms/flatten_tensor.hpp:
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <vector>
+#include <complex>
+#include "openvino/core/node.hpp"
+#include <tensor.hpp>
+#include <host_tensor.hpp>
+
+using namespace ov::op;
+using namespace ov;
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+
+std::vector<std::complex<float>> flatten(const ov::Output<ov::Node>& tensor) {
+    auto host_tensor_ptr = tensor.get_tensor_ptr();
+    auto data_ptr = host_tensor_ptr->get_data_ptr<std::complex<float>>();
+    
+    // Assuming the tensor is 1D or 2D, calculate the total number of elements.
+    auto shape = host_tensor_ptr->get_shape();
+    size_t num_elements = std::accumulate(shape.begin(), shape.end(), 1, std::multiplies<size_t>());
+    
+    // Create a vector from the data pointer.
+    std::vector<std::complex<float>> flat_tensor(data_ptr, data_ptr + num_elements);
+    
+    return flat_tensor;
+}
+
+}  // namespace op
+}  // namespace tensorflow
+}  // namespace frontend
+}  // namespace ov
```
please remove it. Not needed

ysrastogi replied:
Removed

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -24,35 +26,69 @@ namespace op {
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
     auto start = node.get_input(1);
     auto size = node.get_input(2);
-
-    // create axiliary constants
     auto const_one = create_same_type_const_scalar<int32_t>(start, 1);
     auto const_zero = create_same_type_const_scalar<int32_t>(start, 0);
-
     // compute stop values in case non-negative sizes
     auto stop_pos = make_shared<v1::Add>(start, size);
-
     // compute stop values in case negative sizes
     // since TensorFlow supports only -1 among negative sizes
     // assign stop values to the data shape
     Output<Node> stop_neg = make_shared<v3::ShapeOf>(input);
     stop_neg = make_shared<v1::ConvertLike>(stop_neg, size);
-
     // select the correct stop value based on a sign of size value
     auto negative_sizes_mask = make_shared<v1::Less>(size, const_zero);
     // TODO: investigate if we can simplify and replace Select with FloorMod operation
     // like FloorMod(size, input_shape)
     auto stop = make_shared<v1::Select>(negative_sizes_mask, stop_neg, stop_pos);
-
```
please revert this change

ysrastogi replied:
Done

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -12,6 +13,7 @@
 #include "openvino/op/select.hpp"
 #include "openvino/op/shape_of.hpp"
 #include "utils.hpp"
+#include<complex>
```
include complex? unknown header, please remove

ysrastogi replied:
Removed

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -3,7 +3,8 @@
 
 #include "openvino/op/slice.hpp"
-
+#include "helper_ops/flatten_tensor.hpp"
+#include "helper_ops/complex_tensor_reshape.hpp"
```
not needed, revert change

ysrastogi replied:
Done

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Slice.py:
```diff
@@ -27,10 +27,24 @@ def create_slice_net(self, input_shape, input_type, begin_value, size_value):
         dict(input_shape=[10, 5, 1, 5], input_type=tf.float32, begin_value=[5, 1, 0, 3], size_value=[2, 4, -1, -1]),
     ]
 
+    test_data_complex = [
+    dict(input_shape=[6], input_type=tf.complex64, begin_value=[2], size_value=[2]),
+    dict(input_shape=[2, 5, 3], input_type=tf.complex64, begin_value=[0, 1, 0], size_value=[-1, 1, -1]),
+    dict(input_shape=[10, 5, 1, 5], input_type=tf.complex64, begin_value=[5, 1, 0, 3], size_value=[2, 4, -1, -1]),
+]
+
     @pytest.mark.parametrize("params", test_data_basic)
     @pytest.mark.precommit_tf_fe
     @pytest.mark.nightly
     def test_slice_basic(self, params, ie_device, precision, ir_version, temp_dir, use_legacy_frontend):
         self._test(*self.create_slice_net(**params),
                    ie_device, precision, ir_version, temp_dir=temp_dir,
                    use_legacy_frontend=use_legacy_frontend)
+    
+    @pytest.mark.parametrize("params", test_data_complex)
+    @pytest.mark.precommit_tf_fe
+    @pytest.mark.nightly
+    def test_slice_complex(self, params, ie_device, precision, ir_version, temp_dir, use_legacy_frontend):
+        self._test(*self.create_slice_net(**params),
+                   ie_device, precision, ir_version, temp_dir=temp_dir,
+                   use_legacy_frontend=use_legacy_frontend)
```
not correctly implemented layer test.

ysrastogi replied:
How to implement layer test is there any guide?

ysrastogi replied:
Hii @rkazants I have implemented layer tests correctly

rkazants reviewed (changes requested):
please remove unneeded changes and double read instructions how to support complex type in GFI description.

ysrastogi commented:
> please remove unneeded changes and double read instructions how to support complex type in GFI description.

I have removed the unneeded changes and modified slice.cpp according to GFI description.

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -24,37 +26,54 @@ namespace op {
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
     auto start = node.get_input(1);
     auto size = node.get_input(2);
-
-    // create axiliary constants
     auto const_one = create_same_type_const_scalar<int32_t>(start, 1);
     auto const_zero = create_same_type_const_scalar<int32_t>(start, 0);
-
-    // compute stop values in case non-negative sizes
     auto stop_pos = make_shared<v1::Add>(start, size);
-
-    // compute stop values in case negative sizes
-    // since TensorFlow supports only -1 among negative sizes
-    // assign stop values to the data shape
     Output<Node> stop_neg = make_shared<v3::ShapeOf>(input);
     stop_neg = make_shared<v1::ConvertLike>(stop_neg, size);
-
-    // select the correct stop value based on a sign of size value
     auto negative_sizes_mask = make_shared<v1::Less>(size, const_zero);
-    // TODO: investigate if we can simplify and replace Select with FloorMod operation
-    // like FloorMod(size, input_shape)
     auto stop = make_shared<v1::Select>(negative_sizes_mask, stop_neg, stop_pos);
-
-    // broadcast step value
```
please revert these unneeded changes

ysrastogi replied:
I have revert all the unneeded changes

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Slice.py:
```diff
@@ -20,17 +31,45 @@ def create_slice_net(self, input_shape, input_type, begin_value, size_value):
+    def test_slice_complex(self, params, ie_device, precision, ir_version, temp_dir, use_legacy_frontend):
+        self._test(self.create_complex_slice_net(*params),
+                   ie_device, precision, ir_version, temp_dir=temp_dir,
+                   use_legacy_frontend=use_legacy_frontend)
```
please don't change existing layer tests and add separate test suite for complex type. See other PRs with examples of layer tests for complex type.

ysrastogi replied:
Done

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -24,37 +26,54 @@ namespace op {
+    }
 
+    else{
     auto res = make_shared<v8::Slice>(input, start, stop, step);
     set_node_name(node.get_name(), res);
-    return res->outputs();
+    slice = res->outputs();
+    }
+    return slice;
```
please put under `if` only complex type case and leave non-complex type outside `if` after

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -24,10 +26,11 @@ namespace op {
     auto size = node.get_input(2);
 
-    // create axiliary constants
+     // create axiliary constants
```
```suggestion
    // create auxiliary constants
```

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -49,12 +52,37 @@ OutputVector translate_slice_op(const NodeContext& node) {
     auto start_shape = make_shared<v3::ShapeOf>(start);
     auto step = make_shared<v3::Broadcast>(const_one, start_shape);
+    
+
```
```suggestion
```

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -49,12 +52,36 @@ OutputVector translate_slice_op(const NodeContext& node) {
     auto res = make_shared<v8::Slice>(input, start, stop, step);
     set_node_name(node.get_name(), res);
     return res->outputs();
+    
```
```suggestion
```

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -57,4 +83,4 @@ OutputVector translate_slice_op(const NodeContext& node) {
 }  // namespace tensorflow
 }  // namespace frontend
-}  // namespace ov
+}  // namespace ov
```
```suggestion
}  // namespace ov

```

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -24,10 +26,11 @@ namespace op {
 OutputVector translate_slice_op(const NodeContext& node) {
     default_op_checks(node, 3, {"Slice"});
```
```suggestion
    default_op_checks(node, 3, {"Slice"}, true);
```

ysrastogi commented on tests/layer_tests/tensorflow_tests/test_tf_Slice.py:
```diff
@@ -34,3 +33,50 @@ def test_slice_basic(self, params, ie_device, precision, ir_version, temp_dir, u
+    
+
+
+class TestComplexSlice(CommonTFLayerTest):
```
i have run these tests locally.

rkazants commented on src/frontends/tensorflow_common/src/op/slice.cpp:
```diff
@@ -49,6 +52,29 @@ OutputVector translate_slice_op(const NodeContext& node) {
+        auto real_part = make_shared<v8::Slice>(real,start,stop,step);
+        set_node_name(node.get_name(), real_part);
+        auto imag_part = make_shared<v8::Slice>(imag, start, stop, step);
+        set_node_name(node.get_name(), imag_part);
```
please set node name for `concat` operation
